File: app/src/main/java/com/najmi/oreamnos/viewmodel/GenerationState.java
package com.najmi.oreamnos.viewmodel;

/**
 * Immutable state class representing the UI state for content generation.
 * Used by MainViewModel to communicate state changes to MainActivity.
 */
public class GenerationState {

    /**
     * Possible states of the generation process.
     */
    public enum Status {
        /** Initial state, no generation in progress */
        IDLE,
        /** Generation or refinement in progress */
        LOADING,
        /** Generation completed successfully */
        SUCCESS,
        /** Generation failed */
        ERROR
    }

    private final Status status;
    private final String generatedTitle;
    private final String generatedBody;
    private final String sourceCitation;
    private final String errorMessage;
    private final boolean isRefinement;

    private GenerationState(Builder builder) {
        this.status = builder.status;
        this.generatedTitle = builder.generatedTitle;
        this.generatedBody = builder.generatedBody;
        this.sourceCitation = builder.sourceCitation;
        this.errorMessage = builder.errorMessage;
        this.isRefinement = builder.isRefinement;
    }

    // Getters
    public Status getStatus() {
        return status;
    }

    public String getGeneratedTitle() {
        return generatedTitle;
    }

    public String getGeneratedBody() {
        return generatedBody;
    }

    public String getSourceCitation() {
        return sourceCitation;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    public boolean isRefinement() {
        return isRefinement;
    }

    /**
     * Checks if there is generated content available.
     */
    public boolean hasContent() {
        return (generatedTitle != null && !generatedTitle.isEmpty()) ||
                (generatedBody != null && !generatedBody.isEmpty());
    }

    // Factory methods for common states
    public static GenerationState idle() {
        return new Builder().status(Status.IDLE).build();
    }

    public static GenerationState loading(boolean isRefinement) {
        return new Builder()
                .status(Status.LOADING)
                .isRefinement(isRefinement)
                .build();
    }

    public static GenerationState success(String title, String body, String source, boolean isRefinement) {
        return new Builder()
                .status(Status.SUCCESS)
                .generatedTitle(title)
                .generatedBody(body)
                .sourceCitation(source)
                .isRefinement(isRefinement)
                .build();
    }

    public static GenerationState error(String message, boolean isRefinement) {
        return new Builder()
                .status(Status.ERROR)
                .errorMessage(message)
                .isRefinement(isRefinement)
                .build();
    }

    /**
     * Builder pattern for creating GenerationState instances.
     */
    public static class Builder {
        private Status status = Status.IDLE;
        private String generatedTitle = "";
        private String generatedBody = "";
        private String sourceCitation = "";
        private String errorMessage = "";
        private boolean isRefinement = false;

        public Builder status(Status status) {
            this.status = status;
            return this;
        }

        public Builder generatedTitle(String title) {
            this.generatedTitle = title != null ? title : "";
            return this;
        }

        public Builder generatedBody(String body) {
            this.generatedBody = body != null ? body : "";
            return this;
        }

        public Builder sourceCitation(String source) {
            this.sourceCitation = source != null ? source : "";
            return this;
        }

        public Builder errorMessage(String error) {
            this.errorMessage = error != null ? error : "";
            return this;
        }

        public Builder isRefinement(boolean isRefinement) {
            this.isRefinement = isRefinement;
            return this;
        }

        public GenerationState build() {
            return new GenerationState(this);
        }
    }
}

File: app/src/main/java/com/najmi/oreamnos/viewmodel/GenerationState.kt
package com.najmi.oreamnos.viewmodel

/**
 * Sealed class representing the UI state for content generation.
 * Used by MainViewModel to communicate state changes to UI.
 * 
 * Using sealed class instead of enum with builder for better Kotlin pattern matching.
 */
sealed class GenerationState {
    
    /** Initial state, no generation in progress */
    object Idle : GenerationState()
    
    /** Generation or refinement in progress */
    data class Loading(
        @get:JvmName("isRefinementValue")
        val isRefinement: Boolean = false
    ) : GenerationState()
    
    /** Generation completed successfully */
    data class Success(
        @get:JvmName("getTitleValue")
        val generatedTitle: String = "",
        @get:JvmName("getBodyValue")
        val generatedBody: String = "",
        @get:JvmName("getSourceValue")
        val sourceCitation: String = "",
        @get:JvmName("isRefinementValue")
        val isRefinement: Boolean = false
    ) : GenerationState() {
        /** Checks if there is generated content available. */
        override fun hasContent(): Boolean =
            generatedTitle.isNotEmpty() || generatedBody.isNotEmpty()
    }
    
    /** Generation failed */
    data class Error(
        @get:JvmName("getMessageValue")
        val errorMessage: String = "",
        @get:JvmName("isRefinementValue")
        val isRefinement: Boolean = false
    ) : GenerationState()
    
    /** Checks if there is generated content available for any state. */
    open fun hasContent(): Boolean = when (this) {
        is Success -> generatedTitle.isNotEmpty() || generatedBody.isNotEmpty()
        else -> false
    }

    // Java-compatible getter methods (these won't clash now due to @get:JvmName on properties)
    open fun getGeneratedTitle(): String = (this as? Success)?.generatedTitle ?: ""
    open fun getGeneratedBody(): String = (this as? Success)?.generatedBody ?: ""
    open fun getSourceCitation(): String = (this as? Success)?.sourceCitation ?: ""
    open fun getErrorMessage(): String = (this as? Error)?.errorMessage ?: ""
    open fun isRefinement(): Boolean = when (this) {
        is Loading -> isRefinement
        is Success -> isRefinement
        is Error -> isRefinement
        else -> false
    }
    
    companion object {
        /** Factory method for idle state - for Java interop */
        @JvmStatic
        fun idle(): GenerationState = Idle
        
        /** Factory method for loading state */
        @JvmStatic
        fun loading(isRefinement: Boolean): GenerationState = Loading(isRefinement)
        
        /** Factory method for success state */
        @JvmStatic
        fun success(
            title: String?,
            body: String?,
            source: String?,
            isRefinement: Boolean
        ): GenerationState = Success(
            generatedTitle = title ?: "",
            generatedBody = body ?: "",
            sourceCitation = source ?: "",
            isRefinement = isRefinement
        )
        
        /** Factory method for error state */
        @JvmStatic
        fun error(message: String?, isRefinement: Boolean): GenerationState =
            Error(errorMessage = message ?: "", isRefinement = isRefinement)
    }
}

// Extension properties for Kotlin access (optional, kept for Kotlin convenience)
val GenerationState.status: String
    get() = when (this) {
        is GenerationState.Idle -> "IDLE"
        is GenerationState.Loading -> "LOADING"
        is GenerationState.Success -> "SUCCESS"
        is GenerationState.Error -> "ERROR"
    }
